Assignment 3

task 1

Create an array of 7 daily step counts (e.g., from your phone's fitness app) and write code to calculate the mean (average) number of steps.

In [1]:
import numpy as np

steps = np.array([8500, 10200, 6300, 12000, 9800, 7400, 11500])

mean_steps = np.mean(steps)

print("Daily steps:", steps)
print(f"Mean steps: {mean_steps:.2f}")

Daily steps: [ 8500 10200  6300 12000  9800  7400 11500]
Mean steps: 9385.71


**Explanation:**

Mean = sum of all 7 step counts ÷ 7 = 9385.71 steps/day. 
This single number summarizes overall activity level across the week into one representative value. 
Since none of the values are extreme outliers, the mean here is a fair and reliable measure of the user's typical daily steps.

task 2

Given this list of food delivery times in minutes for your last 9 Zomato/Swiggy orders: [32, 28, 29, 45, 30, 31, 60, 30, 29], write a function to find the median delivery time.

In [2]:
import numpy as np

def find_median(times):
    return np.median(times)

delivery_times = np.array([32, 28, 29, 45, 30, 31, 60, 30, 29])
median_time = find_median(delivery_times)

print("Sorted times:", np.sort(delivery_times))
print("Median delivery time:", median_time, "minutes")

Sorted times: [28 29 29 30 30 31 32 45 60]
Median delivery time: 30.0 minutes


**Explanation:**

With 9 (odd) values, the median is simply the middle value once sorted — here, the 5th value out of 9, which is 30 minutes. 
Unlike the mean, the median ignores how far the extreme values (like 45 and 60) are from the center, so it isn't pulled by them. 
This makes 30 minutes a more honest "typical" delivery time than the mean would give, since most orders cluster around 28–32 minutes.

task 3

Imagine you tracked the genres of 12 YouTube videos you watched this week: ['Music', 'Vlog', 'Music', 'Tech', 'Music', 'Vlog', 'Tech', 'Music', 'Comedy', 'Vlog', 'Music', 'Comedy']. Write code to determine the mode (most watched genre).

In [3]:
import pandas as pd

genres = ['Music', 'Vlog', 'Music', 'Tech', 'Music', 'Vlog', 
          'Tech', 'Music', 'Comedy', 'Vlog', 'Music', 'Comedy']

genre_series = pd.Series(genres)
genre_counts = genre_series.value_counts()
mode_genre = genre_series.mode()[0]

print("Genre counts:\n", genre_counts)
print("Most watched genre:", mode_genre)

Genre counts:
 Music     5
Vlog      3
Tech      2
Comedy    2
Name: count, dtype: int64
Most watched genre: Music


**Explanation:**

Mode is the value that occurs most frequently — here, 'Music' appears 5 times out of 12 videos, more than any other genre. 
Mode is the only central tendency measure that works well with categorical (non-numeric) data like genres. 
This tells us Music is clearly the dominant preference this week, followed by Vlogs.

task 4

Take this array of daily UPI transaction amounts (in ₹): [100, 120, 105, 110, 5000, 115, 108]. Calculate both the mean and the median, then explain in 2-3 lines which is a better measure of your 'typical' transaction and why.

**Hint:** Consider the effect of the ₹5000 value.

In [4]:
import numpy as np

transactions = np.array([100, 120, 105, 110, 5000, 115, 108])

mean_amt = np.mean(transactions)
median_amt = np.median(transactions)

print("Transactions:", transactions)
print(f"Mean: ₹{mean_amt:.2f}")
print(f"Median: ₹{median_amt}")

Transactions: [ 100  120  105  110 5000  115  108]
Mean: ₹808.29
Median: ₹110.0


**Explanation:**

The median (₹110) is the better measure of a "typical" transaction here, because the mean (₹808.29) is heavily skewed upward by the single ₹5000 outlier. 
Sorting the data and picking the middle value (median) barely reacts to that one extreme transaction, so it stays close to what most transactions actually look like. 
Whenever data has outliers or is skewed, median is more reliable than mean for describing what's "normal."